File: src/main/java/me/miki/shindo/gui/mainmenu/GuiShindoMainMenu.java
package me.miki.shindo.gui.mainmenu;

import lombok.Getter;
import me.miki.shindo.Shindo;
import me.miki.shindo.gui.mainmenu.impl.*;
import me.miki.shindo.gui.mainmenu.impl.welcome.*;
import me.miki.shindo.management.color.ColorManager;
import me.miki.shindo.management.color.Theme;
import me.miki.shindo.management.color.palette.ColorPalette;
import me.miki.shindo.management.color.palette.ColorType;
import me.miki.shindo.management.event.impl.EventRenderNotification;
import me.miki.shindo.management.language.TranslateText;
import me.miki.shindo.management.nanovg.NanoVGManager;
import me.miki.shindo.management.nanovg.font.Fonts;
import me.miki.shindo.management.nanovg.font.LegacyIcon;
import me.miki.shindo.management.profile.mainmenu.impl.Background;
import me.miki.shindo.management.profile.mainmenu.impl.CustomBackground;
import me.miki.shindo.management.profile.mainmenu.impl.DefaultBackground;
import me.miki.shindo.management.profile.mainmenu.impl.ShaderBackground;
import me.miki.shindo.management.shader.ShaderBackgroundRenderer;
import me.miki.shindo.utils.ColorUtils;
import me.miki.shindo.utils.SkinUtils;
import me.miki.shindo.utils.Sound;
import me.miki.shindo.utils.animation.normal.Animation;
import me.miki.shindo.utils.animation.normal.Direction;
import me.miki.shindo.utils.animation.normal.other.DecelerateAnimation;
import me.miki.shindo.utils.animation.simple.SimpleAnimation;
import me.miki.shindo.utils.mouse.MouseUtils;
import net.minecraft.client.gui.GuiScreen;
import net.minecraft.client.gui.ScaledResolution;
import net.minecraft.client.resources.DefaultPlayerSkin;
import net.minecraft.util.ResourceLocation;
import org.lwjgl.input.Mouse;

import java.awt.*;
import java.util.ArrayList;

public class GuiShindoMainMenu extends GuiScreen {

    private final ArrayList<MainMenuScene> scenes = new ArrayList<MainMenuScene>();

    private final SimpleAnimation skinFocusAnimation = new SimpleAnimation();
    private final SimpleAnimation shopFocusAnimation = new SimpleAnimation();
    private final SimpleAnimation backgroundSelectFocusAnimation = new SimpleAnimation();
    private final SimpleAnimation closeFocusAnimation = new SimpleAnimation();

    private final SimpleAnimation[] backgroundAnimations = new SimpleAnimation[2];

    @Getter
    private MainMenuScene currentScene;


    private boolean soundPlayed = false;


    private Animation fadeIconAnimation, fadeBackgroundAnimation;

    public GuiShindoMainMenu() {

        Shindo instance = Shindo.getInstance();
        boolean firstLogin = instance.getShindoAPI().isFirstLogin();
        ensureDefaultColorScheme(instance, firstLogin);

        for (int i = 0; i < backgroundAnimations.length; i++) {
            backgroundAnimations[i] = new SimpleAnimation();
        }

        scenes.add(new MainScene(this));
        scenes.add(new BackgroundScene(this));
        scenes.add(new ShopScene(this));
        scenes.add(new SkinScene(this));

        scenes.add(new UpdateScene(this));

        scenes.add(new WelcomeMessageScene(this));
        scenes.add(new LanguageSelectScene(this));
        scenes.add(new ThemeSelectScene(this));
        scenes.add(new AccentColorSelectScene(this));
        scenes.add(new CheckingDataScene(this));
        scenes.add(new LastMessageScene(this));


        if (firstLogin) {
            currentScene = getSceneByClass(WelcomeMessageScene.class);
        } else {
            if (instance.getUpdateNeeded()) {
                currentScene = getSceneByClass(UpdateScene.class);
            } else {
                currentScene = getSceneByClass(MainScene.class);
            }
        }
    }

    @Override
    public void initGui() {
        currentScene.initGui();
    }

    @Override
    public void drawScreen(int mouseX, int mouseY, float partialTicks) {

        ScaledResolution sr = new ScaledResolution(mc);

        Shindo instance = Shindo.getInstance();
        NanoVGManager nvg = instance.getNanoVGManager();
        boolean isFirstLogin = instance.getShindoAPI().isFirstLogin();

        backgroundAnimations[0].setAnimation(Mouse.getX(), 16);
        backgroundAnimations[1].setAnimation(Mouse.getY(), 16);

        nvg.setupAndDraw(() -> {

            drawNanoVG(sr, instance, nvg);

            if (!isFirstLogin) {
                drawButtons(mouseX, mouseY, sr, nvg);
            }
        });

        if (currentScene != null) {
            currentScene.drawScreen(mouseX, mouseY, partialTicks);
        }

        if (fadeBackgroundAnimation == null || (fadeBackgroundAnimation != null && !fadeBackgroundAnimation.isDone(Direction.FORWARDS))) {
            nvg.setupAndDraw(() -> drawSplashScreen(sr, nvg));
            if (!soundPlayed) {
                Sound.play("shindo/audio/start.wav", true);
                soundPlayed = true;
            }
        }

        nvg.setupAndDraw(() -> {
            new EventRenderNotification().call();
        });

        super.drawScreen(mouseX, mouseY, partialTicks);
    }

    private void drawNanoVG(ScaledResolution sr, Shindo instance, NanoVGManager nvg) {

        String copyright = "Copyright Mojang AB. Do not distribute!";
        Background currentBackground = instance.getProfileManager().getBackgroundManager().getCurrentBackground();

        if (currentBackground instanceof DefaultBackground) {

            DefaultBackground bg = (DefaultBackground) currentBackground;

            nvg.drawImage(bg.getImage(), -21 + backgroundAnimations[0].getValue() / 90, backgroundAnimations[1].getValue() * -1 / 90, sr.getScaledWidth() + 21, sr.getScaledHeight() + 20);
        } else if (currentBackground instanceof CustomBackground) {

            CustomBackground bg = (CustomBackground) currentBackground;

            nvg.drawImage(bg.getImage(), -21 + backgroundAnimations[0].getValue() / 90, backgroundAnimations[1].getValue() * -1 / 90, sr.getScaledWidth() + 21, sr.getScaledHeight() + 20);
        } else if (currentBackground instanceof ShaderBackground) {

            ShaderBackground bg = (ShaderBackground) currentBackground;

            // Render animated shader background using our renderer
            ShaderBackgroundRenderer.renderShaderBackground(nvg, bg.getShaderFile(), -21 + backgroundAnimations[0].getValue() / 90, backgroundAnimations[1].getValue() * -1 / 90, sr.getScaledWidth() + 21, sr.getScaledHeight() + 20);
        }

        nvg.drawText(copyright, sr.getScaledWidth() - (nvg.getTextWidth(copyright, 9, Fonts.REGULAR)) - 4, sr.getScaledHeight() - 12, new Color(255, 255, 255), 9, Fonts.REGULAR);
        nvg.drawText("Shindo Client v" + instance.getVersion(), 4, sr.getScaledHeight() - 12, new Color(255, 255, 255), 9, Fonts.REGULAR);
    }

    private void drawButtons(int mouseX, int mouseY, ScaledResolution sr, NanoVGManager nvg) {


        Color controlColor = getControlFillColor();

        closeFocusAnimation.setAnimation(MouseUtils.isInside(mouseX, mouseY, sr.getScaledWidth() - 28, 6, 22, 22) ? 1.0F : 0.0F, 16);

        nvg.drawRoundedRect(sr.getScaledWidth() - 28, 6, 22, 22, 4, controlColor);
        nvg.drawCenteredText(LegacyIcon.X, sr.getScaledWidth() - 19F, 8F, new Color(255, 255 - (int) (closeFocusAnimation.getValue() * 200), 255 - (int) (closeFocusAnimation.getValue() * 200)), 18, Fonts.LEGACYICON);

        backgroundSelectFocusAnimation.setAnimation(MouseUtils.isInside(mouseX, mouseY, sr.getScaledWidth() - 28 - 28, 6, 22, 22) ? 1.0F : 0.0F, 16);

        nvg.drawRoundedRect(sr.getScaledWidth() - (28 * 2), 6, 22, 22, 4, controlColor);
        nvg.drawCenteredText(LegacyIcon.IMAGE, sr.getScaledWidth() - (26 * 2) + 6.5F - 1.5F, 9.5F - 1.5F, new Color(255 - (int) (backgroundSelectFocusAnimation.getValue() * 200), 255, 255 - (int) (backgroundSelectFocusAnimation.getValue() * 200)), 18, Fonts.LEGACYICON);

        shopFocusAnimation.setAnimation(MouseUtils.isInside(mouseX, mouseY, sr.getScaledWidth() - (28 * 3), 6, 22, 22) ? 1.0F : 0.0F, 16);

        nvg.drawRoundedRect(sr.getScaledWidth() - (28 * 3), 6, 22, 22, 4, controlColor);
        nvg.drawCenteredText(LegacyIcon.SHOPPING, sr.getScaledWidth() - (26 * 3) + 4.5F, 9.5F, new Color(255 - (int) (shopFocusAnimation.getValue() * 200), 255, 255), 15, Fonts.LEGACYICON);

        skinFocusAnimation.setAnimation(MouseUtils.isInside(mouseX, mouseY, sr.getScaledWidth() - (28 * 4), 6, 22, 22) ? 1.0F : 0.0F, 16);

        nvg.drawRoundedRect(sr.getScaledWidth() - (28 * 4), 6, 22, 22, 4, controlColor);
        nvg.drawCenteredText(LegacyIcon.SKIN, sr.getScaledWidth() - (26 * 4) + 3.5F, 9.5F, new Color(255 - (int) (skinFocusAnimation.getValue() * 200), 255, 255), 15, Fonts.LEGACYICON);

    }

    private void drawSplashScreen(ScaledResolution sr, NanoVGManager nvg) {

        if (fadeIconAnimation == null) {
            fadeIconAnimation = new DecelerateAnimation(100, 1);
            fadeIconAnimation.setDirection(Direction.FORWARDS);
            fadeIconAnimation.reset();
        }

        if (fadeIconAnimation != null) {

            if (fadeIconAnimation.isDone(Direction.FORWARDS) && fadeBackgroundAnimation == null) {
                fadeBackgroundAnimation = new DecelerateAnimation(500, 1);
                fadeBackgroundAnimation.setDirection(Direction.FORWARDS);
                fadeBackgroundAnimation.reset();
            }

            nvg.drawRect(0, 0, sr.getScaledWidth(), sr.getScaledHeight(), new Color(0, 0, 0, fadeBackgroundAnimation != null ? (int) (255 - (fadeBackgroundAnimation.getValue() * 255)) : 255));
            nvg.drawCenteredText(LegacyIcon.SHINDO, sr.getScaledWidth() / 2F, (sr.getScaledHeight() / 2F) - (nvg.getTextHeight(LegacyIcon.SHINDO, 130, Fonts.LEGACYICON) / 2) - 1, new Color(255, 255, 255, (int) (255 - (fadeIconAnimation.getValue() * 255))), 130, Fonts.LEGACYICON);
        }
    }

    @Override
    public void mouseClicked(int mouseX, int mouseY, int mouseButton) {

        ScaledResolution sr = new ScaledResolution(mc);

        boolean isFirstLogin = Shindo.getInstance().getShindoAPI().isFirstLogin();

        if (mouseButton == 0 && !isFirstLogin) {

            if (MouseUtils.isInside(mouseX, mouseY, sr.getScaledWidth() - 28, 6, 22, 22)) {
                mc.shutdown();
            }

            if (MouseUtils.isInside(mouseX, mouseY, sr.getScaledWidth() - (28 * 2), 6, 22, 22) && !this.getCurrentScene().equals(getSceneByClass(BackgroundScene.class))) {
                this.setCurrentScene(this.getSceneByClass(BackgroundScene.class));
            }

            if (MouseUtils.isInside(mouseX, mouseY, sr.getScaledWidth() - (28 * 3), 6, 22, 22)) {
                this.setCurrentScene(this.getSceneByClass(ShopScene.class));
            }

            if (MouseUtils.isInside(mouseX, mouseY, sr.getScaledWidth() - (28 * 4), 6, 22, 22)) {
                this.setCurrentScene(this.getSceneByClass(SkinScene.class));
            }
        }

        currentScene.mouseClicked(mouseX, mouseY, mouseButton);
        super.mouseClicked(mouseX, mouseY, mouseButton);
    }

    @Override
    public void mouseReleased(int mouseX, int mouseY, int mouseButton) {
        currentScene.mouseReleased(mouseX, mouseY, mouseButton);
    }

    @Override
    public void keyTyped(char typedChar, int keyCode) {
        currentScene.keyTyped(typedChar, keyCode);
    }

    @Override
    public void handleInput() {
        super.handleInput();
    }

    @Override
    public void onGuiClosed() {
        currentScene.onGuiClosed();
    }

    public void setCurrentScene(MainMenuScene currentScene) {

        if (this.currentScene != null) {
            this.currentScene.onSceneClosed();
        }

        this.currentScene = currentScene;

        if (this.currentScene != null) {
            this.currentScene.initScene();
        }
    }

    public boolean isDoneBackgroundAnimation() {
        return fadeBackgroundAnimation != null && fadeBackgroundAnimation.isDone(Direction.FORWARDS);
    }

    public MainMenuScene getSceneByClass(Class<? extends MainMenuScene> clazz) {

        for (MainMenuScene s : scenes) {
            if (s.getClass().equals(clazz)) {
                return s;
            }
        }

        return null;
    }

    private void ensureDefaultColorScheme(Shindo instance, boolean forceDefaults) {
        if (instance == null) {
            return;
        }
        ColorManager colorManager = instance.getColorManager();
        if (colorManager == null) {
            return;
        }
        if (forceDefaults || colorManager.getCurrentColor() == null) {
            colorManager.setCurrentColor(colorManager.getColorByName("Default"));
        }
        if (forceDefaults || colorManager.getTheme() == null) {
            colorManager.setTheme(Theme.DARK);
        }
    }

    private Color getControlFillColor() {
        Shindo instance = Shindo.getInstance();
        ColorPalette palette = instance != null && instance.getColorManager() != null ? instance.getColorManager().getPalette() : null;
        Color base = palette != null ? palette.getBackgroundColor(ColorType.NORMAL) : getBackgroundColor();
        return ColorUtils.applyAlpha(base, Math.min(255, base.getAlpha() + 5));
    }

    public Color getBackgroundColor() {
        Shindo instance = Shindo.getInstance();
        ColorPalette palette = instance != null && instance.getColorManager() != null ? instance.getColorManager().getPalette() : null;
        Color base = palette != null ? palette.getBackgroundColor(ColorType.DARK) : new Color(30, 30, 30);
        return ColorUtils.applyAlpha(base, 235);
    }

}

File: src/main/kotlin/me/miki/shindo/gui/mainmenu/GuiShindoMainMenu.kt
package me.miki.shindo.gui.mainmenu

import me.miki.shindo.Shindo
import me.miki.shindo.gui.IShindoScreen
import me.miki.shindo.gui.mainmenu.impl.BackgroundScene
import me.miki.shindo.gui.mainmenu.impl.MainScene
import me.miki.shindo.gui.mainmenu.impl.ShopScene
import me.miki.shindo.gui.mainmenu.impl.SkinScene
import me.miki.shindo.gui.mainmenu.impl.UpdateScene
import me.miki.shindo.gui.mainmenu.impl.welcome.AccentColorSelectScene
import me.miki.shindo.gui.mainmenu.impl.welcome.CheckingDataScene
import me.miki.shindo.gui.mainmenu.impl.welcome.LanguageSelectScene
import me.miki.shindo.gui.mainmenu.impl.welcome.LastMessageScene
import me.miki.shindo.gui.mainmenu.impl.welcome.ThemeSelectScene
import me.miki.shindo.gui.mainmenu.impl.welcome.WelcomeMessageScene
import me.miki.shindo.management.color.ColorManager
import me.miki.shindo.management.color.Theme
import me.miki.shindo.management.color.palette.ColorPalette
import me.miki.shindo.management.color.palette.ColorType
import me.miki.shindo.management.event.impl.EventRenderNotification
import me.miki.shindo.management.nanovg.NanoVGManager
import me.miki.shindo.management.nanovg.font.Fonts
import me.miki.shindo.management.nanovg.font.LegacyIcon
import me.miki.shindo.management.profile.mainmenu.impl.Background
import me.miki.shindo.management.profile.mainmenu.impl.CustomBackground
import me.miki.shindo.management.profile.mainmenu.impl.DefaultBackground
import me.miki.shindo.management.profile.mainmenu.impl.ShaderBackground
import me.miki.shindo.management.shader.ShaderBackgroundRenderer
import me.miki.shindo.utils.ColorUtils
import me.miki.shindo.utils.Sound
import me.miki.shindo.utils.animation.normal.Animation
import me.miki.shindo.utils.animation.normal.Direction
import me.miki.shindo.utils.animation.normal.other.DecelerateAnimation
import me.miki.shindo.utils.animation.simple.SimpleAnimation
import me.miki.shindo.utils.mouse.MouseUtils
import net.minecraft.client.gui.GuiScreen
import net.minecraft.client.gui.ScaledResolution
import org.lwjgl.input.Mouse
import java.awt.Color
import java.util.ArrayList

class GuiShindoMainMenu : GuiScreen(), IShindoScreen {

    private val scenes: ArrayList<MainMenuScene> = ArrayList()

    private val skinFocusAnimation = SimpleAnimation()
    private val shopFocusAnimation = SimpleAnimation()
    private val backgroundSelectFocusAnimation = SimpleAnimation()
    private val closeFocusAnimation = SimpleAnimation()

    private val backgroundAnimations = Array(2) { SimpleAnimation() }

    var currentScene: MainMenuScene? = null
        private set

    private var soundPlayed = false

    private var fadeIconAnimation: Animation? = null
    private var fadeBackgroundAnimation: Animation? = null

    init {
        val instance = Shindo.getInstance()
        val firstLogin = instance.shindoAPI.isFirstLogin()
        ensureDefaultColorScheme(instance, firstLogin)

        scenes.add(MainScene(this))
        scenes.add(BackgroundScene(this))
        scenes.add(ShopScene(this))
        scenes.add(SkinScene(this))
        scenes.add(UpdateScene(this))
        scenes.add(WelcomeMessageScene(this))
        scenes.add(LanguageSelectScene(this))
        scenes.add(ThemeSelectScene(this))
        scenes.add(AccentColorSelectScene(this))
        scenes.add(CheckingDataScene(this))
        scenes.add(LastMessageScene(this))

        currentScene = if (firstLogin) {
            getSceneByClass(WelcomeMessageScene::class.java)
        } else {
            if (instance.updateNeeded) {
                getSceneByClass(UpdateScene::class.java)
            } else {
                getSceneByClass(MainScene::class.java)
            }
        }
    }

    override fun initGui() {
        currentScene?.initGui()
    }

    override fun drawScreen(mouseX: Int, mouseY: Int, partialTicks: Float) {
        val sr = ScaledResolution(mc)

        val instance = Shindo.getInstance()
        val nvg = instance.nanoVGManager
        val isFirstLogin = instance.shindoAPI.isFirstLogin()

        backgroundAnimations[0].setAnimation(Mouse.getX().toFloat(), 16.0)
        backgroundAnimations[1].setAnimation(Mouse.getY().toFloat(), 16.0)

        nvg!!.setupAndDraw(Runnable {
            drawNanoVG(sr, instance, nvg)

            if (!isFirstLogin) {
                drawButtons(mouseX, mouseY, sr, nvg)
            }
        })

        currentScene?.drawScreen(mouseX, mouseY, partialTicks)

        if (fadeBackgroundAnimation == null || (fadeBackgroundAnimation != null && !fadeBackgroundAnimation!!.isDone(Direction.FORWARDS))) {
            nvg.setupAndDraw(Runnable { drawSplashScreen(sr, nvg) })
            if (!soundPlayed) {
                Sound.play("shindo/audio/start.wav", true)
                soundPlayed = true
            }
        }

        nvg.setupAndDraw(Runnable { EventRenderNotification().call() })

        super.drawScreen(mouseX, mouseY, partialTicks)
    }

    private fun drawNanoVG(sr: ScaledResolution, instance: Shindo, nvg: NanoVGManager) {
        val copyright = "Copyright Mojang AB. Do not distribute!"
        val currentBackground = instance.profileManager.backgroundManager.currentBackground

        when (currentBackground) {
            is DefaultBackground -> {
                nvg.drawImage(
                    currentBackground.image,
                    -21f + backgroundAnimations[0].value / 90,
                    backgroundAnimations[1].value * -1 / 90,
                    sr.scaledWidth + 21f,
                    sr.scaledHeight + 20f
                )
            }
            is CustomBackground -> {
                nvg.drawImage(
                    currentBackground.image,
                    -21f + backgroundAnimations[0].value / 90,
                    backgroundAnimations[1].value * -1 / 90,
                    sr.scaledWidth + 21f,
                    sr.scaledHeight + 20f
                )
            }
            is ShaderBackground -> {
                ShaderBackgroundRenderer.renderShaderBackground(
                    nvg,
                    currentBackground.shaderFile,
                    -21f + backgroundAnimations[0].value / 90,
                    backgroundAnimations[1].value * -1 / 90,
                    sr.scaledWidth + 21f,
                    sr.scaledHeight + 20f
                )
            }
        }

        nvg.drawText(copyright, sr.scaledWidth - (nvg.getTextWidth(copyright, 9f, Fonts.REGULAR)) - 4, sr.scaledHeight - 12f, Color.WHITE, 9f, Fonts.REGULAR)
        nvg.drawText("Shindo Client v" + instance.version, 4f, sr.scaledHeight - 12f, Color.WHITE, 9f, Fonts.REGULAR)
    }

    private fun drawButtons(mouseX: Int, mouseY: Int, sr: ScaledResolution, nvg: NanoVGManager) {
        val controlColor = getControlFillColor()

        closeFocusAnimation.setAnimation(
            if (MouseUtils.isInside(mouseX, mouseY, sr.scaledWidth - 28f, 6f, 22f, 22f)) 1.0f else 0.0f,
            16.0
        )

        nvg.drawRoundedRect(sr.scaledWidth - 28f, 6f, 22f, 22f, 4f, controlColor)
        nvg.drawCenteredText(
            LegacyIcon.X,
            sr.scaledWidth - 19f,
            8f,
            Color(255, 255 - (closeFocusAnimation.value * 200).toInt(), 255 - (closeFocusAnimation.value * 200).toInt()),
            18f,
            Fonts.LEGACYICON
        )

        backgroundSelectFocusAnimation.setAnimation(
            if (MouseUtils.isInside(mouseX, mouseY, sr.scaledWidth - 56f, 6f, 22f, 22f)) 1.0f else 0.0f,
            16
        )

        nvg.drawRoundedRect(sr.scaledWidth - 56f, 6f, 22f, 22f, 4f, controlColor)
        nvg.drawCenteredText(
            LegacyIcon.IMAGE,
            sr.scaledWidth - 52f + 6.5f - 1.5f,
            9.5f - 1.5f,
            Color(255 - (backgroundSelectFocusAnimation.value * 200).toInt(), 255, 255 - (backgroundSelectFocusAnimation.value * 200).toInt()),
            18f,
            Fonts.LEGACYICON
        )

        shopFocusAnimation.setAnimation(
            if (MouseUtils.isInside(mouseX, mouseY, sr.scaledWidth - 84f, 6f, 22f, 22f)) 1.0f else 0.0f,
            16
        )

        nvg.drawRoundedRect(sr.scaledWidth - 84f, 6f, 22f, 22f, 4f, controlColor)
        nvg.drawCenteredText(
            LegacyIcon.SHOPPING,
            sr.scaledWidth - 78f + 4.5f,
            9.5f,
            Color(255 - (shopFocusAnimation.value * 200).toInt(), 255, 255),
            15f,
            Fonts.LEGACYICON
        )

        skinFocusAnimation.setAnimation(
            if (MouseUtils.isInside(mouseX, mouseY, sr.scaledWidth - 112f, 6f, 22f, 22f)) 1.0f else 0.0f,
            16
        )

        nvg.drawRoundedRect(sr.scaledWidth - 112f, 6f, 22f, 22f, 4f, controlColor)
        nvg.drawCenteredText(
            LegacyIcon.SKIN,
            sr.scaledWidth - 104f + 3.5f,
            9.5f,
            Color(255 - (skinFocusAnimation.value * 200).toInt(), 255, 255),
            15f,
            Fonts.LEGACYICON
        )
    }

    private fun drawSplashScreen(sr: ScaledResolution, nvg: NanoVGManager) {
        if (fadeIconAnimation == null) {
            fadeIconAnimation = DecelerateAnimation(100, 1.0)
            fadeIconAnimation!!.setDirection(Direction.FORWARDS)
            fadeIconAnimation!!.reset()
        }

        if (fadeIconAnimation != null) {
            if (fadeIconAnimation!!.isDone(Direction.FORWARDS) && fadeBackgroundAnimation == null) {
                fadeBackgroundAnimation = DecelerateAnimation(500, 1.0)
                fadeBackgroundAnimation!!.setDirection(Direction.FORWARDS)
                fadeBackgroundAnimation!!.reset()
            }

            nvg.drawRect(0f, 0f, sr.scaledWidth.toFloat(), sr.scaledHeight.toFloat(), Color(0, 0, 0, if (fadeBackgroundAnimation != null) (255 - (fadeBackgroundAnimation!!.getValue() * 255)).toInt() else 255))
            nvg.drawCenteredText(
                LegacyIcon.SHINDO,
                sr.scaledWidth / 2f,
                (sr.scaledHeight / 2f) - (nvg.getTextHeight(LegacyIcon.SHINDO, 130f, Fonts.LEGACYICON) / 2) - 1,
                Color(255, 255, 255, (255 - (fadeIconAnimation!!.getValue() * 255)).toInt()),
                130f,
                Fonts.LEGACYICON
            )
        }
    }

    override fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        val sr = ScaledResolution(mc)

        val isFirstLogin = Shindo.getInstance().shindoAPI.isFirstLogin()

        if (mouseButton == 0 && !isFirstLogin) {
            if (MouseUtils.isInside(mouseX, mouseY, sr.scaledWidth - 28f, 6f, 22f, 22f)) {
                mc.shutdown()
            }

            if (MouseUtils.isInside(mouseX, mouseY, sr.scaledWidth - 56f, 6f, 22f, 22f) && currentScene != getSceneByClass(BackgroundScene::class.java)) {
                setCurrentScene(getSceneByClass(BackgroundScene::class.java))
            }

            if (MouseUtils.isInside(mouseX, mouseY, sr.scaledWidth - 84f, 6f, 22f, 22f)) {
                setCurrentScene(getSceneByClass(ShopScene::class.java))
            }

            if (MouseUtils.isInside(mouseX, mouseY, sr.scaledWidth - 112f, 6f, 22f, 22f)) {
                setCurrentScene(getSceneByClass(SkinScene::class.java))
            }
        }

        currentScene?.mouseClicked(mouseX, mouseY, mouseButton)
        super.mouseClicked(mouseX, mouseY, mouseButton)
    }

    override fun mouseReleased(mouseX: Int, mouseY: Int, mouseButton: Int) {
        currentScene?.mouseReleased(mouseX, mouseY, mouseButton)
    }

    override fun keyTyped(typedChar: Char, keyCode: Int) {
        currentScene?.keyTyped(typedChar, keyCode)
    }

    override fun handleInput() {
        super.handleInput()
    }

    override fun onGuiClosed() {
        currentScene?.onGuiClosed()
    }

    fun setCurrentScene(currentScene: MainMenuScene?) {
        this.currentScene?.onSceneClosed()
        this.currentScene = currentScene
        this.currentScene?.initScene()
    }

    fun isDoneBackgroundAnimation(): Boolean {
        return fadeBackgroundAnimation != null && fadeBackgroundAnimation!!.isDone(Direction.FORWARDS)
    }

    fun getSceneByClass(clazz: Class<out MainMenuScene>): MainMenuScene? {
        for (scene in scenes) {
            if (scene.javaClass == clazz) {
                return scene
            }
        }
        return null
    }

    private fun ensureDefaultColorScheme(instance: Shindo?, forceDefaults: Boolean) {
        if (instance == null) {
            return
        }
        val colorManager: ColorManager? = instance.colorManager
        if (colorManager == null) {
            return
        }
        if (forceDefaults || colorManager.currentColor == null) {
            colorManager.currentColor = colorManager.getColorByName("Default")
        }
        if (forceDefaults || colorManager.theme == null) {
            colorManager.theme = Theme.DARK
        }
    }

    private fun getControlFillColor(): Color {
        val instance = Shindo.getInstance()
        val palette: ColorPalette? = if (instance != null && instance.colorManager != null) instance.colorManager.palette else null
        val base = palette?.getBackgroundColor(ColorType.NORMAL) ?: getBackgroundColor()
        return ColorUtils.applyAlpha(base, Math.min(255, base.alpha + 5))
    }

    fun getBackgroundColor(): Color {
        val instance = Shindo.getInstance()
        val palette: ColorPalette? = if (instance != null && instance.colorManager != null) instance.colorManager.palette else null
        val base = palette?.getBackgroundColor(ColorType.DARK) ?: Color(30, 30, 30)
        return ColorUtils.applyAlpha(base, 235)
    }
}
